### PR TITLE
[Feature] Add tracking data structures and data flow

### PR DESCRIPTION
**Pull request by JingweiZhang12**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add BaseVideoDataset, MOTChallengeDataset, TrackDataSample,  ImgQuotaSampler and TrackDataPreprocessor.
## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by JingweiZhang12** [comment]
@GT9505, please review this PR.

File: mmdet/datasets/base_video_dataset.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+from mmengine.dataset import BaseDataset
+from mmengine.fileio import FileClient
+from mmengine.logging import print_log
+
+from mmdet.datasets.api_wrappers import COCO
+from mmdet.registry import DATASETS
+
+
+@DATASETS.register_module()
+class BaseVideoDataset(BaseDataset):
```
Comment (BIGWangYuDong) [inline mmdet/datasets/base_video_dataset.py:16]: Seems BaseVideoDataset is force set coco style as default. Can we rename it to `CocoVideoDataset`, and add a pure BaseVideoDataset?

File: configs/_base_/datasets/mot_challenge.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        ]),
+    dict(type='PackTrackInputs')
+]
+
+# dataloader
+train_dataloader = dict(
+    batch_size=2,
+    num_workers=2,
+    persistent_workers=True,
+    # batch_sampler=dict(type='AspectRatioBatchSampler'), # used in image dataset: like LVIS # noqa: E501
```
Comment (RangiLyu) [inline configs/_base_/datasets/mot_challenge.py:60]: Add more comments to tell users this dataset is a video based dataset so we do not use AspectRatioBatchSampler

File: mmdet/datasets/samplers/img_quota_sampler.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            assert not self.test_mode, "'ClassBalancedDataset' should not "
+            'exist in test mode'
+            video_indices = self.dataset.repeat_indices
+            for index in video_indices:
+                self.indices.extend([(index, frame_ind) for frame_ind in range(
+                    ori_dataset.get_len_per_video(index))])
+        else:
+            assert isinstance(
+                self.dataset,
+                BaseVideoDataset), f'{type(self.dataset)} is not supported'
```
Comment (RangiLyu) [inline mmdet/datasets/samplers/img_quota_sampler.py:68]: the error message should be more clear: `f'ImgQuotaSampler only supports video dataset, but got {type(self.dataset)} '`

File: mmdet/engine/runner/loops.py [hunk 4]
```diff
@@ -37,3 +40,104 @@ def run(self):
...
+                self.video_outputs = track_data_sample.clone()
+            else:
+                self.video_outputs.video_data_samples.append(
+                    track_data_sample[0].clone())
+                self.video_outputs.set_metainfo(
+                    dict(video_length=self.video_outputs.video_length + 1))
+                if frame_id == ori_video_len - 1:
+                    self.evaluator.process(
+                        data_samples=[self.video_outputs],
+                        data_batch=data_batch)
```
Comment (RangiLyu) [inline mmdet/engine/runner/loops.py:137]: Can you move this part to the evaluator or metric?

File: configs/_base_/datasets/mot_challenge.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        filter_key_img=True),
+    dict(
+        type='TransformBroadcaster',
+        share_random_params=True,
+        transforms=[
+            dict(type='LoadImageFromFile'),
+            dict(type='LoadTrackAnnotations', with_instance_id=True),
+            dict(
+                type='RandomResize',
+                scale=(1088, 1088),
```
Comment (hhaAndroid) [inline configs/_base_/datasets/mot_challenge.py:20]: (1088, 1088) shape to the front, easy to unify the user to modify

File: mmdet/structures/track_data_sample.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+from typing import List, Optional, Sequence
+
+import numpy as np
+import torch
+from mmengine.structures import BaseDataElement
+
+from .det_data_sample import DetDataSample
+
+
+class TrackDataSample(BaseDataElement):
```
Comment (RangiLyu) [inline mmdet/structures/track_data_sample.py:11]: This datasample is designed only for tracking or for all video tasks?

File: configs/_base_/datasets/mot_challenge.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                ratio_range=(0.8, 1.2),
+                keep_ratio=True,
+                clip_object_border=False),
+            dict(type='PhotoMetricDistortion')
+        ]),
+    dict(
+        type='TransformBroadcaster',
+        share_random_params=False,
```
Comment (hhaAndroid) [inline configs/_base_/datasets/mot_challenge.py:28]: Is there a basis for setting this parameter to true or false? Suggest to add a comment

File: mmdet/datasets/base_video_dataset.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+        self.cat_ids = self.coco.get_cat_ids(
+            cat_names=self.metainfo['classes'])
+        self.cat2label = {cat_id: i for i, cat_id in enumerate(self.cat_ids)}
+        self.cat_img_map = copy.deepcopy(self.coco.cat_img_map)
+        # used in `filter_data`
+        self.img_ids_with_ann = set()
+
+        img_ids = self.coco.get_img_ids()
+        total_ann_ids = []
+        single_video_id = 100000
```
Comment (hhaAndroid) [inline mmdet/datasets/base_video_dataset.py:47]: Suggest adding a comment

File: mmdet/datasets/base_video_dataset.py [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+            if ann.get('instance_id', None):
+                instance['instance_id'] = ann['instance_id']
+            else:
+                # image dataset usually has no `instance_id`.
+                # Therefore, we set it to `i`.
+                instance['instance_id'] = i
+            if len(instance) > 0:
+                instances.append(instance)
+        data_info['instances'] = instances
+        return data_info
```
Comment (hhaAndroid) [inline mmdet/datasets/base_video_dataset.py:146]: We should assert if the number of instances is 0. If it is 0, an error should be notified

File: mmdet/datasets/samplers/img_quota_sampler.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+            number should be identical across all processes in the distributed
+            group. Default: 0.
+    """
+
+    def __init__(self, dataset: Sized, seed: int = 0) -> None:
+        rank, world_size = get_dist_info()
+        self.rank = rank
+        self.world_size = world_size
+        self.epoch = 0
+        self.seed = seed if seed is not None else 0
```
Comment (hhaAndroid) [inline mmdet/datasets/samplers/img_quota_sampler.py:32]: Forcing 0 doesn't quite make sense. We want to ensure full diversity when no seed is provided, so a different seed should be randomized when no seed is provided.  defaults to None is better

File: mmdet/datasets/samplers/img_quota_sampler.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+        self.dataset = dataset
+        self.indices = []
+        # Hard code here to handle different dataset wrapper
+        # TODO: refactor this part
```
Comment (hhaAndroid) [inline mmdet/datasets/samplers/img_quota_sampler.py:37]: This is really not good, is there a refactoring plan?

File: mmdet/models/data_preprocessors/track_data_preprocessor.py [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+import torch.nn.functional as F
+from mmengine.model import BaseDataPreprocessor
+
+from mmdet.registry import MODELS
+from mmdet.structures import TrackDataSample
+from mmdet.structures.mask import BitmapMasks
+
+
+@MODELS.register_module()
+class TrackDataPreprocessor(BaseDataPreprocessor):
```
Comment (hhaAndroid) [inline mmdet/models/data_preprocessors/track_data_preprocessor.py:17]: Can inherit DetDataPreprocessor

File: mmdet/datasets/base_video_dataset.py (hunk 1 above)
Comment (JingweiZhang12) [reply mmdet/datasets/base_video_dataset.py:16]: All the ann files in tracking are coco style after the dataset conversion.

File: mmdet/structures/track_data_sample.py (hunk 6 above)
Comment (JingweiZhang12) [reply mmdet/structures/track_data_sample.py:11]: It's only used in tracking task

File: mmdet/datasets/base_video_dataset.py (hunk 9 above)
Comment (JingweiZhang12) [reply mmdet/datasets/base_video_dataset.py:146]: But the length of  `instances` is 0 in testset and it's not an error.

File: mmdet/models/data_preprocessors/track_data_preprocessor.py (hunk 12 above)
Comment (JingweiZhang12) [reply mmdet/models/data_preprocessors/track_data_preprocessor.py:17]: If inherit DetDataPreprocessor, TrackDataPreprocessor will have some unnecessary property, such as `gt_instances`, and `pred_instances`. This is not in line with the function of `TrackDataPreprocessor`.

File: mmdet/engine/runner/loops.py (hunk 4 above)
Comment (JingweiZhang12) [reply mmdet/engine/runner/loops.py:137]: Sounds better. We will support both video predictions and image predicts in the Metric and deprecates these video loops.

File: mmdet/datasets/base_video_dataset.py (hunk 9 above)
Comment (hhaAndroid) [reply mmdet/datasets/base_video_dataset.py:146]: We can use the `test_mode` parameter to determine

File: mmdet/engine/runner/loops.py (hunk 4 above)
Comment (RangiLyu) [reply mmdet/engine/runner/loops.py:137]: You can leave a TODO here to avoid forgetting.

**Review by RangiLyu — APPROVED** [review]
LGTM

File: mmdet/models/data_preprocessors/track_data_preprocessor.py [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+                    det_data_sample = track_data_sample[i]
+                    masks = det_data_sample.gt_instances.masks
+                    # TODO: whether to use BitmapMasks
+                    assert isinstance(masks, BitmapMasks)
+                    batch_input_shape = det_data_sample.batch_input_shape
+                    det_data_sample.gt_instances.masks = masks.pad(
+                        batch_input_shape, pad_val=self.mask_pad_value)
+
+
+def stack_batch(tensors: List[torch.Tensor],
```
Comment (hhaAndroid) [inline mmdet/models/data_preprocessors/track_data_preprocessor.py:198]: Is it possible to simply extend mmengine's s-function to support it? If so, it is recommended to add a TODO note for subsequent integration into mmengine.

File: mmdet/engine/runner/loops.py (hunk 4 above)
Comment (JingweiZhang12) [reply mmdet/engine/runner/loops.py]: Have removed that part of code.

File: mmdet/datasets/base_video_dataset.py [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+        Returns:
+            Any: Depends on ``self.pipeline``.
+        """
+        if isinstance(idx, tuple):
+            assert len(idx) == 2, 'The length of idx must be 2: '
+            '(video_index, frame_index)'
+            video_idx, frame_idx = idx[0], idx[1]
+        else:
+            video_idx, frame_idx = idx, None
+
```
[the remark this answers is not in the document]
Comment (JingweiZhang12) [reply mmdet/datasets/base_video_dataset.py:230]: Yes. In training mode, `ImgQuotaSampler` is optional and only used in some methods, like QDTrack, to align the precision on the small MOT datasets.

File: mmdet/datasets/samplers/img_quota_sampler.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+from mmengine.dataset import ClassBalancedDataset, ConcatDataset
+from mmengine.dist import get_dist_info, sync_random_seed
+from torch.utils.data import Sampler
+
+from mmdet.registry import DATA_SAMPLERS
+from ..base_video_dataset import BaseVideoDataset
+
+
+@DATA_SAMPLERS.register_module()
+class ImgQuotaSampler(Sampler):
```
[the remark this answers is not in the document]
Comment (JingweiZhang12) [reply mmdet/datasets/samplers/img_quota_sampler.py:16]: I rename it as TrackImgSampler which accounts for two key meaning: tracking and images. Also, I add a more detailed docstring about this sampler.

File: mmdet/structures/track_data_sample.py [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+    @video_data_samples.deleter
+    def video_data_samples(self):
+        del self._video_data_samples
+
+    def __getitem__(self, index):
+        assert hasattr(self,
+                       '_video_data_samples'), 'video_data_samples not set'
+        return self._video_data_samples[index]
+
+    def get_key_frames(self):
```
[the remark this answers is not in the document]
Comment (JingweiZhang12) [reply mmdet/structures/track_data_sample.py:132]: I still think `get_key_frames` is more neat and clear enough since the DataSample only contains annotation info, no other data, such as image tensor.
